### PR TITLE
Bug 1211028: commiting only changed resources.

### PR DESCRIPTION
**Pull request by jotes**
Hi @mathjazz, could you review?

### CODE REVIEW

**Comment by mathjazz** [comment]
We can't rely on `RepositorySyncLog`, because we sometimes trigger sync manually, e.g. when adding a new locale to a project.

I also don't think filtering translations with `date__gte=latest_sync` is necessary, because we already take care of that using `Entity.has_changed`.

**Comment by jotes** [comment]
@mathjazz Hi, could you look again?

File: pontoon/sync/changeset.py [hunk 1]
```diff
@@ -44,13 +44,15 @@ def __init__(self, db_project, vcs_project, now):
         self.commit_authors_per_locale = defaultdict(list)
         self.locales_to_commit = set()
 
-    def update_vcs_entity(self, locale, db_entity, vcs_entity):
+    def update_vcs_entity(self, locale, db_entity, vcs_entity, force=False):
```
Comment (mathjazz) [inline pontoon/sync/changeset.py]: Do you ever use `force`?

File: pontoon/sync/tests/test_changeset.py [hunk 2]
```diff
@@ -317,6 +317,63 @@ def test_obsolete_db(self):
...
+
+        with patch.object(self.main_db_entity, 'has_changed', return_value=False) as mock_has_changed:
+            resource_file = MagicMock()
+            self.changeset.update_vcs_entity(self.translated_locale, self.main_db_entity, MagicMock())
+            self.changeset.vcs_project.resources = {
+                self.main_db_entity.resource.path: resource_file
+            }
+
+            self.changeset.execute_update_vcs()
+
```
Comment (mathjazz) [inline pontoon/sync/tests/test_changeset.py]: Remove blank line.

File: pontoon/sync/tests/test_changeset.py [hunk 3]
```diff
@@ -317,6 +317,63 @@ def test_obsolete_db(self):
...
+            self.changeset.execute_update_vcs()
+            assert mock_has_changed.called
+            assert resource_file.save.called
+
+    def test_unchanged_resources_sync(self):
+        """
+        Synchronization shouldn't modify resources if their entities weren't changed.
+        """
+
```
Comment (mathjazz) [inline pontoon/sync/tests/test_changeset.py]: Remove blank line.

File: pontoon/sync/tests/test_changeset.py [hunk 4]
```diff
@@ -385,3 +442,18 @@ def test_multiple_translations(self):
 
         assert_equal(self.changeset.commit_authors_per_locale[self.translated_locale.code],
             [first_author])
+
+
+    def test_no_translations(self):
+        """
+        We don't attribute anyone if there aren't any new translations.
+        """
+
```
Comment (mathjazz) [inline pontoon/sync/tests/test_changeset.py]: Remove blank line.

File: pontoon/sync/tests/test_changeset.py [hunk 5]
```diff
@@ -317,6 +317,63 @@ def test_obsolete_db(self):
...
+        TranslationFactory.create(locale=self.translated_locale, entity=self.main_db_entity,
+            approved=True, date=aware_datetime(2015, 1, 1))
+        resource_file = MagicMock()
+        self.changeset.vcs_project.resources = {
+            self.main_db_entity.resource.path: resource_file
+        }
+
+        with patch.object(self.main_db_entity, 'has_changed', return_value=False) as mock_has_changed:
+            self.changeset.update_vcs_entity(self.translated_locale, self.main_db_entity, MagicMock())
```
Comment (mathjazz) [inline pontoon/sync/tests/test_changeset.py]: Add bank line.

File: pontoon/sync/tests/test_changeset.py [hunk 6]
```diff
@@ -317,6 +317,63 @@ def test_obsolete_db(self):
...
+
+    def test_changed_resources_sync(self):
+        """
+        Synchronization should modify resource files only if there are changed translations.
+        """
+        TranslationFactory.create(locale=self.translated_locale, entity=self.main_db_entity,
+            approved=True, date=aware_datetime(2015, 1, 1))
+
+        resource_file = MagicMock()
+
```
Comment (mathjazz) [inline pontoon/sync/tests/test_changeset.py]: Remove blank line.

File: pontoon/sync/tests/test_changeset.py [hunk 7]
```diff
@@ -317,6 +317,63 @@ def test_obsolete_db(self):
...
+    def test_unchanged_resources_sync(self):
+        """
+        Synchronization shouldn't modify resources if their entities weren't changed.
+        """
+
+        TranslationFactory.create(locale=self.translated_locale, entity=self.main_db_entity,
+            approved=True, date=aware_datetime(2015, 1, 1))
+        resource_file = MagicMock()
+        self.changeset.vcs_project.resources = {
```
Comment (mathjazz) [inline pontoon/sync/tests/test_changeset.py]: Add blank line.

File: pontoon/sync/tests/test_changeset.py [hunk 8]
```diff
@@ -317,6 +317,63 @@ def test_obsolete_db(self):
...
+            assert resource_file.save.called
+
+    def test_unchanged_resources_sync(self):
+        """
+        Synchronization shouldn't modify resources if their entities weren't changed.
+        """
+
+        TranslationFactory.create(locale=self.translated_locale, entity=self.main_db_entity,
+            approved=True, date=aware_datetime(2015, 1, 1))
+        resource_file = MagicMock()
```
Comment (mathjazz) [inline pontoon/sync/tests/test_changeset.py]: Add blank line.

File: pontoon/sync/tests/test_changeset.py [hunk 9]
```diff
@@ -317,6 +317,63 @@ def test_obsolete_db(self):
...
+            assert mock_has_changed.called
+            assert not resource_file.save.called
+
+    def test_changed_resources_sync(self):
+        """
+        Synchronization should modify resource files only if there are changed translations.
+        """
+        TranslationFactory.create(locale=self.translated_locale, entity=self.main_db_entity,
+            approved=True, date=aware_datetime(2015, 1, 1))
+
```
Comment (mathjazz) [inline pontoon/sync/tests/test_changeset.py]: Remove blank line.

**Comment by mathjazz** [comment]
Nice work, r+wc. I'd remove that `force` unsused flag.

I also added pointed out some nits WRT blank lines in tests, which are totally optional to fix.

**Comment by jotes** [comment]
@mathjazz hi, could you look again?

**Comment by mathjazz** [comment]
A thing of beauty. Please rebase as single commit, using "Fix bug 1211028" at the begining of commit message.

**Comment by jotes** [comment]
Done :)

**Pull request merged**